py-utils/src/utils/validator/v_pkg.py

```
@@ -16,67 +16,127 @@
-				result = self.__search_pkg("pip3 list")
-			if result.find(f"{pkg}") == -1:
-				raise VError(errno.EINVAL,
-					     "pip3 pkg: %s not found" % pkg)
+    """Pkg related validations."""
+
+    def __init__(self):
+        self.passwdless_ssh_enabled = False
+        self.ssh = None
+
+    def __execute_cmd(self, cmd):
+        """
+        Execute command using SSHChannel or SimpleProcees
+        Uses SSHChannel to execute ommand on host if passwordless ssh is not configured.
+        Uses SimpleProcess to execute command if no host is provided in args or
+        passwordless ssh is configured on host.
+        """
+        # print(f"Running {cmd}")
+        if self.ssh:
+            retcode, result = self.ssh.execute(cmd)
+        else:
+            handler = SimpleProcess(cmd)
+            stdout, stderr, retcode = handler.run()
+            result = stdout.decode("utf-8") if retcode ==0 else stderr.decode("utf-8")
+        if retcode != 0:
+            raise VError(
+                errno.EINVAL,
+                "Command failure. cmd: %s stderr: %s" %(cmd, result))
+        return result
+
+    def validate(self, v_type: str, args: list, host: str = None):
+        """
+        Process rpm validations.
+        Usage (arguments to be provided):
+        1. pkg validate_rpms host (optional) [packagenames]
+        2. pkg validate_pip3s host (optional) [pip3 packagenames]
+        """
+        if host:
+            try:
+                # Ensure we can perform passwordless ssh and there are no prompts
+                NetworkV().validate('passwordless',
+                    [pwd.getpwuid(os.getuid()).pw_name, host])
+                self.passwdless_ssh_enabled = True
+            except VError:
+                # Create ssh connection in case passwordless ssh is not configured
+                host_details = re.search(r"(\w+):(.+)@(.+):(\d+)", host)
+                if not host_details or len(host_details.groups()) != 4:
+                    raise VError(
```

There is no need to catch and rethrow error unless you plan to catch some other exception. 
Moreover the error sent by earlier exception is lost here because you dont catch it properly

---

When passwordless ssh is not configured on that host network validator throws VError. Here we catch that specific exception purposefully and validate the host url in exception block.
Let me make it more clear.

---

As not catching none other exception, I removed `except VError` and now its generic.

---

Again the reason to catch and rethrow is not clear. 
except: --> Issue. Exception information is lost. 
except VError: --> Issue. Exception information is lost. 
except DivideByZero as e: OK. exception information is in 'e'

---

As discussed, I'm proceeding with passwordless ssh validation approach and removed host_url approach.
Hence the previous version is reverted.

---

Code is committed with some identified use cases now.